rgw/beast: drop privileges after binding ports

emmericp opened the pull request:
Fixes: http://tracker.ceph.com/issues/36041

Signed-off-by: Paul Emmerich <paul.emmerich@croit.io>

- [x] References tracker ticket
- [x] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug (ideas how to test this? couldn't find any place where the privilege dropping is being tested at the moment)

cbodley commented:
thanks! i made some changes to the rgw/verify suite to run radosgw on privileged ports at https://github.com/cbodley/ceph/commits/wip-qa-rgw-priv-port. i'll test your pr against that

emmericp commented:
jenkins, re-test this please

cbodley commented on src/rgw/rgw_asio_frontend.cc:
```diff
@@ -350,6 +351,23 @@ int AsioFrontend::init()
+  }
+  if (uid && gid) {
+      ldout(ctx(), 1) << "set uid:gid to " << uid << ":" << gid << " (" << uid_string << ":" << gid_string << ")" << dendl;
+  }
```
since this code is essentially duplicated from `global_init()`, can we find a way to move both into a common function? maybe something like `int CephContext::drop_privileges(std::ostream&)`?

cbodley commented:
passed testing in http://pulpito.ceph.com/cbodley-2018-09-25_17:00:24-rgw-wip-36041-distro-basic-smithi/ (though lots of noise from unrelated failures)

emmericp commented:
I've refactored it as requested. I've also ported civetweb to use the new drop_privileges(), this ensures that you get the proper log messages about the dropped privileges on startup (it currently doesn't log the user change)

I've renamed the commit since it's now mostly about the stuff in init and the fixed bug is only a few lines of code.

cbodley commented on src/rgw/rgw_civetweb_frontend.cc:
```diff
@@ -77,13 +77,6 @@ int RGWCivetWebFrontend::run()
-  if (! uid_string.empty()) {
-    conf_map.emplace("run_as_user", std::move(uid_string));
-  }
-
```
`mg_start()` is doing several things after its call to `set_uid_option()` - and while it doesn't look like any of it depends on the new uid/gid, that could change upstream in the future. so without better test coverage of its options, i'd prefer not to change how the civetweb side works

emmericp commented:
Reverted it to the first version but with drop_privileges as a separate function

cbodley commented:
@emmericp looks great, thanks
